### PR TITLE
Add s3 transaction store support

### PR DESCRIPTION
**Pull request by Gisson**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,93 @@
+
```
Comment (brunotm) [inline .gitignore:1]: remove this file please

File: store/s3/s3.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+		})
+}
+
+type Store struct {
+	data       *s3.S3
+	bucketName string
+	prefix     string
+}
+
+// URI in the form of s3://<access>:<secret>/<bucket>/<path>
```
Comment (brunotm) [inline store/s3/s3.go]: Add Go style comments so it get documented properly out of the box.

File: store/s3/s3.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+func New(uri string) (*Store, error) {
+	//TODO
```
Comment (brunotm) [inline store/s3/s3.go]: Which todo?

File: store/s3/s3.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	var err error
+	u, err := url.Parse(uri)
+	if err != nil {
+		return nil, err
```
Comment (brunotm) [inline store/s3/s3.go]: wrap the error with the %w directive for fmt.Errorf so the callee knows where it came from and can also compare the underlying error.

File: store/s3/s3.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	}
+	if u.Scheme != "s3" {
+		return nil, fmt.Errorf("Invalid uri scheme for s3")
+	}
+	var awsconfig *aws.Config = aws.NewConfig()
+	var creds *credentials.Credentials
+	var sess *session.Session
+	secretkey, haspassword := u.User.Password()
```
Comment (brunotm) [inline store/s3/s3.go]: name variables using Go conventions: secretKey, hasPassword

File: store/s3/s3.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+		creds = credentials.NewStaticCredentialsFromCreds(credentials.Value{
+			AccessKeyID:     u.User.Username(),
+			SecretAccessKey: secretkey,
+		})
+		awsconfig.WithCredentials(creds)
+	}
+
+	if err != nil {
+		return nil, err
+	}
```
Comment (brunotm) [inline store/s3/s3.go]: empty line for readability after this one :)

File: store/s3/s3.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+	if reg, ok := u.Query()["region"]; ok {
+		sess, err = session.NewSession(awsconfig, aws.NewConfig().WithCredentials(credentials.NewEnvCredentials()), aws.NewConfig().WithRegion(reg[0]))
```
Comment (brunotm) [inline store/s3/s3.go]: We are only loading from env and dropping the user provided parameters.

File: store/s3/s3.go [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+	} else {
+		sess, err = session.NewSession(awsconfig, aws.NewConfig().WithCredentials(credentials.NewEnvCredentials()))
```
Comment (brunotm) [inline store/s3/s3.go]: same as above.

**Review by brunotm — COMMENTED** [review]
The on going implementation looks good :) Let's move forward with it.

Notes:
- Use gofmt on the code
- Add empty lines to improve readbility
- Don't load environment variables from library code
- Wrap errors when returning before crossing the package boundaries
- Open a new MR with the url.Parse as suggested :)

File: .gitignore (hunk 1 above)
Comment (Gisson) [reply .gitignore:1]: This file is needed in the repository, we can however add it in another MR/

File: store/s3/s3.go [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+
+// New function creates a new storage object with a subtype of s3.
+// receives uri which should be in the form of s3://<access>:<secret>/<bucket>/<prefix>.
+// returns a store with subtype s3.
+func New(uri string) (*Store, error) {
+	var err error
+	u, err := url.Parse(uri)
+
+	if err != nil {
+		return nil, fmt.Errorf("%w", err)
```
Comment (brunotm) [inline store/s3/s3.go]: `return nil, fmt.Errorf("store/s3: failed to parse uri: %w", err)`

File: store/s3/s3.go [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+	if hasPassword { // If access and secret key is used, then initialize it
+		creds = credentials.NewStaticCredentialsFromCreds(credentials.Value{
+			AccessKeyID:     u.User.Username(),
+			SecretAccessKey: secretKey,
+		})
+		awsconfig.WithCredentials(creds)
+	}
+
+	if err != nil {
```
Comment (brunotm) [inline store/s3/s3.go]: We are not checking any error here. Can be removed.

File: store/s3/s3.go [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+// returns an error in case of failing to delete the object.
+func (s *Store) Delete(name string) error {
+	input := &s3.DeleteObjectInput{
+		Bucket: aws.String(s.bucketName),
+		Key:    aws.String(s.prefix + "/" + name),
+	}
+	_, err := s.data.DeleteObject(input)
+
+	if err != nil {
+		return fmt.Errorf("%w", err)
```
Comment (brunotm) [inline store/s3/s3.go]: return fmt.Errorf("store/s3: error deleting object: %w", err)

File: store/s3/s3.go [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+	}
+
+	return config, nil
+}
+
+// Has function return true or false for either the object exists or not in the s3 bucket.
+// receives a string with the path to the object.
+// returns a boolean (true if object exists, false otherwise) and an error in case of unexpected behaviour.
+func (s *Store) Has(name string) (exists bool, err error) {
+
```
Comment (brunotm) [inline store/s3/s3.go:139]: Can't Has() just reuse Get() and check for errors?

File: store/s3/s3.go [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+func (s *Store) Iter(callback func(name string, config transaction.Config) (proceed bool)) (err error) {
+	listObjectInput := &s3.ListObjectsInput{
+		Bucket: aws.String(s.bucketName),
+		Prefix: aws.String(s.prefix),
+	}
+
+	outputs, err := s.data.ListObjects(listObjectInput)
+
+	if err != nil {
+		return fmt.Errorf("%w", err)
```
Comment (brunotm) [inline store/s3/s3.go]: return fmt.Errorf("store/s3: failed to list bucket: %w", err)

File: .gitignore [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+# Output of the go coverage tool, specifically when used with LiteIDE
+*.out
+
+# Dependency directories (remove the comment below to include it)
+# vendor/
+
+### Go Patch ###
+/vendor/
+/Godeps/
+
```
Comment (brunotm) [inline .gitignore:22]: Remove those lines as we are keeping the vendored deps for build reliability, security and speed.

File: .gitignore (hunk 1 above)
Comment (brunotm) [reply .gitignore:1]: We can keep as long as we remove the vendor lines.

**Review by brunotm — COMMENTED** [review]
Looks good. A few nits to fix, simple tests and we are good to go.

File: store/s3/s3.go [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+
+func init() {
+	store.Register("s3",
+		func(uri string) (s store.Store, err error) {
+			return New(uri)
+		})
+}
+
+type Store struct {
+	data       *s3.S3 // S3 data source object.
```
Comment (brunotm) [inline store/s3/s3.go]: switch this to s3iface.S3API to make testing possible by mocking the client.

File: store/s3/s3.go [hunk 16]
```diff
@@ -39,11 +41,11 @@ func New(uri string) (*Store, error) {
 	u, err := url.Parse(uri)
 
 	if err != nil {
-		return nil, fmt.Errorf("%w", err)
+		return nil, wrapError(fmt.Sprintf("failed to parse URI %s", uri), err)
 	}
 
 	if u.Scheme != "s3" {
-		return nil, fmt.Errorf("Invalid uri scheme for s3")
+		return nil, wrapError(fmt.Sprintf("invalid S3 scheme in URI %s. Scheme is %s", uri, u.Scheme), fmt.Errorf("Invalid S3 URI"))
```
Comment (brunotm) [inline store/s3/s3.go]: yeah... this wrapError doesn't make any sense.
fmt.Errorf could do all in a single pass and more efficiently.

**Pull request merged**